src/bin/pg_autoctl/supervisor.c

```diff
@@ -282,21 +604,52 @@ supervisor_loop(Supervisor *supervisor)
 }
 
 
+/*
+ * supervisor_find_recently_removed_dynamic is responsible for finding a
+ * recently removed dynamic service in the dynamicRecentlyRemoved array. If the
+ * service if found, then it is removed from this array.
+ */
+static bool
+supervisor_find_recently_removed_dynamic(Supervisor *supervisor, pid_t pid,
+										 Service **result)
+{
+	Service *recentlyRemoved = supervisor->dynamicRecentlyRemoved.services;
+	int serviceCount = supervisor->dynamicRecentlyRemoved.serviceCount;
+	int serviceIndex;
+
+	for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+	{
+		Service *service = &recentlyRemoved[serviceIndex];
+		if (service->pid == pid)
+		{
+			*result = service;
+
+			/* Now that the service is found, remove it from the array */
+			recentlyRemoved[serviceIndex] = recentlyRemoved[serviceCount - 1];
+			supervisor->dynamicRecentlyRemoved.serviceCount--;
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
 /*
  * supervisor_find_service loops over the SubProcess array to find given pid and
  * return its entry in the array.
  */
 static bool
 supervisor_find_service(Supervisor *supervisor, pid_t pid, Service **result)
 {
-	int serviceCount = supervisor->serviceCount;
-	int serviceIndex = 0;
+	Supervisor_it sit;
+	Service *service;
 
-	for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+	foreach_supervised_service(&sit, supervisor, service)
```

Comment:
As said before I preferred the previous coding. This macro does not look like an improvement to me, and it blurs the memory aspect of things.